Add support for local login

sanderDijkxhoorn opened the pull request:
- You can now save your bike credentials to a .json file
- Only shows non api features if there is no api credentials saved in the local storage
- Login using your credentials file
- Fix controls-test error (someone forgot to add batteryChargingLevel
- Supports “Bikey” & “VanMoof key exporter”

sanderDijkxhoorn committed:
Add support for local login

- You can now save your bike credentials to a .json file
- Only shows non api features if there is no api credentials saved in the local storage
- Login using your credentials file
- Fix controls-test error (someone forgot to add batteryChargingLevel

sanderDijkxhoorn committed:
Hopefully fix save info on iOS

sanderDijkxhoorn committed:
Open in new tab for iOS Bluefy for now

sanderDijkxhoorn committed:
Plz just something?

rjackson commented:
Just came over to the repo to see if anybody had been working on this! Would've been interested in having a go myself :)

Personally, I feel like having to wrangle a JSON file might be a bit awkward for most people – and indeed appears to be awkward in the code, especially re: the Bluefy issues you'd been having.

Instead, I'd recommend accepting `encryptionKey` and `passcode` explicitly. This ensures compatibility for any users who've extracted their data via:

- @grossartig's [VanMoof Encryption Key Exporter](https://vanoof.grossartig.io/)
- Maybe Cowboy's [Bikey app](https://bikey-app.cowboy.bike/), if that exposes those details (it only stores them locally, so users would need a way to migrate credentials to new phones if VanMoof servers ever went down)

We could provide a  JSON import / export method for those who have saved the JSON file either from _VanMoof Encryption Key Exporter_ (saves `data.data.bikeDetails`, or `BikeDetails[]`) or from Mooovy (in whatever JSON format we deem appropriate), but I would consider that a bonus convenience feature – not the primary means of authentication.

sanderDijkxhoorn commented:
> Just came over to the repo to see if anybody had been working on this! Would've been interested in having a go myself :)
> 
> Personally, I feel like having to wrangle a JSON file might be a bit awkward for most people – and indeed appears to be awkward in the code, especially re: the Bluefy issues you'd been having.
> 
> Instead, I'd recommend accepting `encryptionKey` and `passcode` explicitly. This ensures compatibility for any users who've extracted their data via:
> 
> * @grossartig's [VanMoof Encryption Key Exporter](https://vanoof.grossartig.io/)
> * Maybe Cowboy's [Bikey app](https://bikey-app.cowboy.bike/), if that exposes those details (it only stores them locally, so users would need a way to migrate credentials to new phones if VanMoof servers ever went down)
> 
> We could provide a JSON import / export method for those who have saved the JSON file either from _VanMoof Encryption Key Exporter_ (saves `data.data.bikeDetails`, or `BikeDetails[]`) or from Mooovy (in whatever JSON format we deem appropriate), but I would consider that a bonus convenience feature – not the primary means of authentication.

With Bikey you have to use a JSON file too.

Without the other data in the file we can’t determine which type of bike the user is using and you can only use one bike at a time. 

Saving the file is almost as user friendly as possible if you’d ask me, except for the Bluefy issue which should be able to be resolved I just didn’t have enough time to take a good look at it. 

The user just has to select the file and they are logged in. 

I could add a additional login option so you can login with just the keys.

rjackson commented:
> I could add a additional login option so you can login with just the keys.

I think that would be really useful.

I take the point about only being able to use one bike at a time, but I'm conscious that the VanMoof Encryption Key Exporter is making the rounds and its initial web & print layout only highlights a few specific details (frame number, mac address, passcode, encryption key, user key id). It'd be really useful if we can allow users who have printed off those details, and skipped or lost the JSON file, to get back into their bike.

mjarkk commented on components/Login.tsx:
```diff
@@ -23,6 +23,8 @@ export default function Login({ setCredentials }: LoginArgs) {
         password: '',
     })
 
+    const [loginMethod, setLoginMethod] = useState('email')
```
```suggestion
    const [loginMethod, setLoginMethod] = useState<'email' | 'file'>('email')
```

mjarkk commented on components/Login.tsx:
```diff
@@ -53,39 +55,131 @@ export default function Login({ setCredentials }: LoginArgs) {
+                </div>
+                <FormError error={error} />
+            </form>
+            <div className='loginMethod'>
```
Wat dacht je van zoiets?
Zodra je dan op _"Login using JSON file"_ klikt opened die dan direct het select file modal en na het selecteren van een file gebruikt deze direct om in te loggen (gebruiker hoeft dan ook niet meer te klikken op login)

<img width="513" alt="image" src="https://github.com/mjarkk/vanmoof-web-controller/assets/15320763/cf5a0492-824c-4ab6-820a-386f85e10477">

sanderDijkxhoorn replied:
Ja, dat is een betere oplossing 👍🏻

mjarkk commented on components/Controls.tsx:
```diff
@@ -9,11 +9,29 @@ import { compareVersions } from 'compare-versions'
 export default function BikeControls({ bike, api, disconnect }: BikeControlsArgs) {
+    // Check if api is available
+    if (!api) {
+        console.error('Api not available')
```
This code path is not an error thus this console.error can be removed

sanderDijkxhoorn replied:
Was for debugging but I forgot to remove it

mjarkk commented on components/Connect.tsx:
```diff
@@ -70,6 +70,46 @@ export default function BluetoothConnect({ bikeCredentials, setBikeInstance, bac
+                    }}
+                    secondary
+                >
+                    Save login info
```
Maybe `backup login credentials` is a better sentence here?

mjarkk commented on components/Connect.tsx:
```diff
@@ -70,6 +70,46 @@ export default function BluetoothConnect({ bikeCredentials, setBikeInstance, bac
+
+                            pre.appendChild(code)
+
+                            div.appendChild(pre)
```
We are using react there is no need to create html elements by hand.
You can easily add a state to this component that contains the credentials and only show the credentials if it contains content.

mjarkk commented on components/Connect.tsx:
```diff
@@ -70,6 +70,46 @@ export default function BluetoothConnect({ bikeCredentials, setBikeInstance, bac
+
+                        // If the user is on iOS and has the Bluefy browser, we can't save the file directly
+                        // so we'll just show the user the JSON file and they can save it manually
+                        if (navigator.userAgent.includes('iPhone') && navigator.userAgent.includes('Bluefy')) {
```
This feels a bit hacky.
How about always showing the credentials on screen when clicking this button and also downloading them.
That way the user will see a file is downloaded thus not needing to copy the contents on screen and for browsers that for some wired reason can't download files they can copy the contents on screen.

sanderDijkxhoorn replied:
It is pretty big especially for mobile devices hence I preferred to only showing it when necessarily but will do till we come up with a better idea maybe a like box that can be minimised and with just a copy button

mjarkk replied:
You can make the font size smaller and place it within a box with scrollbars maybe?
And add a button to copy the credentials to the clipboard maybe?

sanderDijkxhoorn replied:
There is a copy button already?

mjarkk commented:
I was also thinking what kind of key does (bikey)[https://bikey-app.cowboy.com] and [vanoof.grossartig.io](https://vanoof.grossartig.io) export?
Maybe we should use their file format so all the vm tools are compatible with each other.

sanderDijkxhoorn commented:
> I was also thinking what kind of key does (bikey)[https://bikey-app.cowboy.com] and [vanoof.grossartig.io](https://vanoof.grossartig.io) export?
> Maybe we should use their file format so all the vm tools are compatible with each other.

Cowboy is just using the result of the api request the one with ?includebike or something like that. Not sure about that website.

mjarkk commented on components/Connect.tsx:
```diff
@@ -70,6 +70,46 @@ export default function BluetoothConnect({ bikeCredentials, setBikeInstance, bac
+            <P vertical={10}>
+                <Button
+                    onClick={() => {
+                        const blob = new Blob([JSON.stringify(bikeCredentials)], { type: 'application/json' })
```
A fancy json file is much better for these kinds of things.
_Adding `, null, 4` to a `JSON.stringify` will print a fancy JSON file_
```suggestion
                        const blob = new Blob([JSON.stringify(bikeCredentials, null, 4)], { type: 'application/json' })
```

sanderDijkxhoorn committed:
Fancy JSON

Co-authored-by: Mark Kopenga <mkopenga@gmail.com>

sanderDijkxhoorn commented on components/Login.tsx:
```diff
@@ -53,39 +55,131 @@ export default function Login({ setCredentials }: LoginArgs) {
+                }
+
+                // Parse the JSON as BikeCredentials (There could be one bike but also multiple)
+                const bikes = JSON.parse(reader.result as string) as Array<BikeCredentials>
```
> You are now using `as Array<BikeCredentials>` but this does not mean the underlaying data will be that type.
> 
> I think we should validate the contents of the json file if it matches what we want.

I tried to validate it but no luck, could you help with that?

mjarkk replied:
You can use zod for this:

Introduction to zod: [Zod + Generics is HEAVEN (YouTube)](https://youtu.be/9N50YV5NHaE)
zod documentation: [zod.dev](https://zod.dev)

sanderDijkxhoorn replied:
I am using zod now but I don't like to redeclare everything in the interface I tried to use the interface type and parse it using that but it won't accept it: 'BikeCredentials' only refers to a type, but is being used as a value here.

sanderDijkxhoorn committed:
Update components/Login.tsx

Co-authored-by: Mark Kopenga <mkopenga@gmail.com>

sanderDijkxhoorn commented on components/Controls.tsx:
```diff
@@ -9,11 +9,29 @@ import { compareVersions } from 'compare-versions'
 }
 
 export default function BikeControls({ bike, api, disconnect }: BikeControlsArgs) {
+    // Check if api is available
```
> I would prefer this code: This way you don't have the same code twice
> 
> ```tsx
> return (
>         <BikeContext.Provider value={bike}>
>             <ApiContext.Provider value={api}>
>                 <BikeStats bike={bike} />
>                 <SpeedLimit bike={bike} />
>                 <PowerLevel bike={bike} />
>                 <BellTone bike={bike} />
>                 <SoundBoard />
>                 {api ?
>                     <>
>                         <ShareBike bike={bike} api={api} />
>                         <CurrentShares bike={bike} api={api} />
>                     </>
>                     : undefined}
>                 <Button onClick={disconnect} secondary>
>                     Disconnect bike
>                 </Button>
>             </ApiContext.Provider>
>         </BikeContext.Provider>
>     )
> ```

Actually that won't work because of the ApiContext provider

sanderDijkxhoorn replied:
What do you think about this?
```dart
    const bikeStats = (
        <>
            <BikeStats bike={bike} />
            <SpeedLimit bike={bike} />
            <PowerLevel bike={bike} />
            <BellTone bike={bike} />
            <SoundBoard />
            {api ?
                <>
                    <ShareBike bike={bike} api={api!} />
                    <CurrentShares bike={bike} api={api!} />
                </>
                : undefined}
            <Button onClick={disconnect} secondary>Disconnect bike</Button>
        </>
    )

    return (
        <BikeContext.Provider value={bike}>
            {api ? (
                <ApiContext.Provider value={api!}>
                    {bikeStats}
                </ApiContext.Provider>
            ) : (
                bikeStats
            )}
        </BikeContext.Provider>
    )

mjarkk replied:
Oh wait I see my example above didn't work :),
Maybe something like this?:

```tsx
return (
        <BikeContext.Provider value={bike}>
                <BikeStats bike={bike} />
                <SpeedLimit bike={bike} />
                <PowerLevel bike={bike} />
                <BellTone bike={bike} />
                <SoundBoard />
                {api ?
                    <ApiContext.Provider value={api}>
                        <ShareBike bike={bike} api={api} />
                        <CurrentShares bike={bike} api={api} />
                    </ApiContext.Provider >
                    : undefined}
                <Button onClick={disconnect} secondary>
                    Disconnect bike
                </Button>
        </BikeContext.Provider>
    )
```

sanderDijkxhoorn replied:
> Oh wait I see my example above didn't work :), Maybe something like this?:
> 
> ```tsx
> return (
>         <BikeContext.Provider value={bike}>
>                 <BikeStats bike={bike} />
>                 <SpeedLimit bike={bike} />
>                 <PowerLevel bike={bike} />
>                 <BellTone bike={bike} />
>                 <SoundBoard />
>                 {api ?
>                     <ApiContext.Provider value={api}>
>                         <ShareBike bike={bike} api={api} />
>                         <CurrentShares bike={bike} api={api} />
>                     </ApiContext.Provider >
>                     : undefined}
>                 <Button onClick={disconnect} secondary>
>                     Disconnect bike
>                 </Button>
>         </BikeContext.Provider>
>     )
> ```

I changed it into this now, will test if everything still functions later

sanderDijkxhoorn replied:
Works 👍🏻

sanderDijkxhoorn committed:
Improvements

Using react for the backup key parts, could still use some improvements.

Simplify the login using a file part.

Upgrade caniuse-lite.

Adept the controls test page to the changes.

sanderDijkxhoorn committed:
Simplify controls & additional support

You can now login using exports from "VanMoof Encryption Key Exporter" & from "Bikey"

sanderDijkxhoorn committed:
Make the backup key text smaller

mjarkk commented on components/Connect.tsx:
```diff
@@ -70,6 +84,42 @@ export default function BluetoothConnect({ bikeCredentials, setBikeInstance, bac
             </P>
+
+            {/* If Backup button pressed hide it */}
+            {!ShowBackupKey && (
+                <P vertical={10}>
+                    <Button
+                        onClick={() => setShowBackupKey(true)}
+                        secondary
+                    >
+                        Backup login credentials
+                    </Button>
+                </P>
+            )}
+
+            {/* If Backup button pressed show the key */}
+            {ShowBackupKey && (
+                <>
+                    {/* Copy button */}
+                    <Button onClick={handleCopy} secondary>
+                        Copy credentials to clipboard
+                    </Button>
+
+                    <P vertical={10}>
+                        {/* Download button */}
+                        <Button onClick={handleDownload} secondary>
+                            Download credentials as file
+                        </Button>
+                    </P>
+
+                    {/* Show the key */}
+                    <textarea
+                        readOnly
+                        value={parsedCredentials}
+                        style={{ width: '85%', height: '200px', fontSize: '0.5rem' }}
+                    />
+                </>
+            )}
```
This can become an inline if else

```suggestion
            {ShowBackupKey ? (
                <div>
                    {/* If Backup button pressed show the key */}

                    <Button onClick={handleCopy} secondary>
                        Copy credentials to clipboard
                    </Button>

                    <P vertical={10}>
                        <Button onClick={handleDownload} secondary>
                            Download credentials as file
                        </Button>
                    </P>

                    <textarea
                        readOnly
                        value={parsedCredentials}
                        style={{ width: '85%', height: '200px', fontSize: '0.5rem' }}
                    />
                </div>
            ) : (
                <P vertical={10}>
                    {/* If Backup button pressed hide it */}

                    <Button
                        onClick={() => setShowBackupKey(true)}
                        secondary
                    >
                        Backup login credentials
                    </Button>
                </P>
            )}
```

mjarkk commented on components/Login.tsx:
```diff
@@ -53,39 +56,148 @@ export default function Login({ setCredentials }: LoginArgs) {
+                        primary: z.string(),
+                        secondary: z.string(),
+                    })),
+                    links: z.nullable(z.object({
```
I think [nullish](https://zod.dev/?id=nullish) is a better property for this

mjarkk commented:
I cannot test this code atm and the bikey app is gone from the app stores.
Because of that i will close this pr for now.
There is a pr open for manually adding bikes using a mac and encryption key.. see: https://github.com/mjarkk/vanmoof-web-controller/pull/55

sanderDijkxhoorn commented:
> I cannot test this code atm and the bikey app is gone from the app stores.
> Because of that i will close this pr for now.
> There is a pr open for manually adding bikes using a mac and encryption key.. see: #55

https://apps.apple.com/nl/app/bikey-app/id6451117309?l=en-GB
The bikey app is still up on the AppStore^

mjarkk commented:
Ah i looked at the the playstore link here: https://play.google.com/store/apps/details?id=com.cowboy.bikeyapp
